extensions/amp-video/1.0/video-wrapper.js

```diff
@@ -0,0 +1,207 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import * as Preact from '../../../src/preact';
+import {MIN_VISIBILITY_RATIO_FOR_AUTOPLAY} from '../../../src/video-interface';
+import {useEffect, useRef, useState} from '../../../src/preact';
+import {useResourcesNotify} from '../../../src/preact/utils';
+// Source for this constant is css/video-autoplay.css
+import {Deferred} from '../../../src/utils/promise';
+import {
+  EMPTY_METADATA,
+  parseFavicon,
+  parseOgImage,
+  parseSchemaImage,
+  setMediaSession,
+} from '../../../src/mediasession-helper';
+import {cssText as autoplayCss} from '../../../build/video-autoplay.css';
+
+/**
+ * @param {!VideoWrapperProps} props
+ * @return {PreactDef.Renderable}
+ */
+export function VideoWrapper({
+  component: Component,
+  autoplay = false,
+  controls = false,
+  noaudio = false,
+  mediasession = true,
+  children,
+  // Prevent override in passthroughProps.
+  muted: unusedMuted,
+  ...passthroughProps
+}) {
+  useResourcesNotify();
+
+  const [muted, setMuted] = useState(autoplay);
+  const [metadata, setMetadata] = useState(null);
+  const [playing, setPlaying] = useState(false);
+  const [userInteracted, setUserInteracted] = useState(false);
+
+  const wrapperRef = useRef(null);
+  const playerRef = useRef(null);
+
+  const readyDeferredRef = useRef(new Deferred());
+
+  const play = () => {
+    readyDeferredRef.current.promise.then(() => {
+      playerRef.current.play();
+    });
+  };
+
+  const pause = () => {
+    readyDeferredRef.current.promise.then(() => {
+      playerRef.current.pause();
+    });
+  };
+
+  useEffect(() => {
+    if (!mediasession || !metadata || !playing) {
+      return;
+    }
+    setMediaSession(window, metadata, play, pause);
+  }, [mediasession, playing, metadata]);
+
+  return (
+    <div
+      ref={wrapperRef}
+      style={{position: 'relative', width: '100%', height: '100%'}}
+    >
+      <Component
+        ref={playerRef}
+        // Ensure that props controlled here are excluded from passthroughProps,
+        // see destructuring statement above.
+        muted={muted}
+        controls={controls && (!autoplay || userInteracted)}
+        onLoad={() => readyDeferredRef.current.resolve()}
+        onLoadedMetadata={() => {
+          if (!mediasession) {
+            return;
+          }
+
+          const metadata = playerRef.current.getMetadata() || {
```

Comment:
Can we use the event target instead here?

Reply:
Should we use a known API on `HTMLMediaElement` for this or some other event data instead?

Reply:
That's a good question. Partially why I brought up the event target is because I wanted us to deal with event and target APIs sooner. But I think let's defer this for now. Do you have a video tracker already? If not, can you create and we add a TODO to it? See the [tracker for fit-text](https://github.com/ampproject/amphtml/issues/28281) for an example. But to answer a bigger question: overall we'd want to stay as close to the `HTMLMediaElement` as we can.